cpp/src/arrow/array-test.cc

```diff
@@ -1764,4 +1765,169 @@ TEST(TestRechunkArraysConsistently, Plain) {
   }
 }
 
+class ConcatenateTest : public ::testing::Test {
+ protected:
+  ConcatenateTest()
+      : rng_(seed_),
+        sizes_({0, 1, 2, 4, 16, 31, 1234}),
+        null_probabilities_({0.0, 0.1, 0.5, 0.9, 1.0}) {}
```

Comment:
You could consider parameterizing the test case over `null_probabilities_`

Reply:
This PR originally used a parameterized test, but the multiple properties I'm sweeping over made the [parameters excessively complicated](https://github.com/apache/arrow/pull/3746#discussion_r260038384)